src/commands/command_runner.py

```diff
@@ -0,0 +1,187 @@
+
+        if len(command) <= 0:
+            raise Exception("Command content can't be empty")
+
+        self.__command_id = command_id
+        self.__command: list = shlex.split(command)
+        self.__on_line: OnLineCallback = on_line
+        self.__on_exit: OnExitCallback = on_exit
+
+        self.__process_id: Optional[int] = None
+        self.__process: Optional[subprocess.Popen] = None
+        self.__status: CommandStatus = CommandStatus.NotStarted
+        self.__output: Optional[CommandFinalResult] = None
+
+    @property
+    def command_id(self):
+        return self.__command_id
+
+    @property
+    def command(self):
+        return self.__command
+
+    @property
+    def process_id(self):
+        return self.__process_id
+
+    @property
+    def status(self):
+        return self.__status
+
+    @property
+    def output(self):
+        return self.__output
+
+    @property
+    def is_complete(self) -> bool:
+        """ :return: whether the process completed (both successfully or killed) """
+
+        return self.__status == CommandStatus.Killed or self.__status == CommandStatus.FinishedSuccessfully
+
+    def run(self) -> threading.Thread:
+        """ Runs the command in a separate thread
+        While Running:
+        - status is set to Running
+        - on_line is called for each line output except empty lines
+        When done:
+        - status is set to FinishedSuccessfully
+        - on_exit is called with the return code and std output
+        :return: the thread where the process was running
+        """
+
+        def run_in_thread():
+            process = subprocess.Popen(self.__command,
+                                       shell=True,
+                                       stdout=subprocess.PIPE,
+                                       stderr=subprocess.STDOUT,
+                                       creationflags=subprocess.CREATE_NEW_PROCESS_GROUP)
+
+            self.__status = CommandStatus.Running
+
+            self.__process = process
+            self.__process_id = process.pid
+
+            if self.__on_line is not None:
+                for line in process.stdout:
+                    should_pass, clean_line = verify_and_clean_line(line)
+                    if should_pass:
+                        self.__on_line(CommandLineOutput(self.__command_id, clean_line))
+
+            process.wait()
+
+            output, _ = self.__process.communicate()
+
+            self.__finish(output, CommandStatus.FinishedSuccessfully)
+
+            return
+
+        thread = threading.Thread(target=run_in_thread)
+        thread.start()
+
+        return thread
+
+    def kill(self):
+        """ Kills the process and each subprocess launched by the process
+        Also sets status to Killed
+        Still fills output with output of command and calls on_exit
+        """
+
+        if self.__status.Running and self.__process_id:
+            ps_util_process = psutil.Process(self.__process_id)
```

Comment:
We should check if ps_util_process is null, if its null as a result we should sent result code 100+ ( the os result codes are from 0 to 20~something i think we can send 100+ result codes as application specific error. For example code 104: You are tying to kill process that is not running

Reply:
Correct, I just checked, and [psutil raises `NoSuchProcess`](https://psutil.readthedocs.io/en/latest/#process-class) if the pid doesn't exist.

Reply:
I added the `TODO`s so we can fix this together here https://github.com/NHristovski/ServerHelper-ClientApp/pull/1/commits/fff65d099efc0e2086ce0997f1cf1f1186016596.